File: mobile_app/lib/models/generation.dart
import 'package:flutter/material.dart';
import 'package:freezed_annotation/freezed_annotation.dart';

part 'generation.freezed.dart';
part 'generation.g.dart';

/// Main generation model tracking AI pipeline progress and results
@freezed
class Generation with _$Generation {
  const factory Generation({
    required String id,
    required String profileId,
    required String jobId,
    required DocumentType documentType,
    required GenerationStatus status,
    required GenerationProgress progress,
    GenerationResult? result,
    String? errorMessage,
    @Default(0) int tokensUsed,
    double? generationTime,
    required DateTime createdAt,
    DateTime? completedAt,
    DateTime? startedAt,
    DateTime? updatedAt,
    String? estimatedCompletion,
  }) = _Generation;

  factory Generation.fromJson(Map<String, dynamic> json) =>
      _$GenerationFromJson(json);
}

/// Pagination metadata for generation lists
@freezed
class Pagination with _$Pagination {
  const factory Pagination({
    required int total,
    required int limit,
    required int offset,
    required bool hasNext,
    required bool hasPrevious,
  }) = _Pagination;

  factory Pagination.fromJson(Map<String, dynamic> json) =>
      _$PaginationFromJson(json);
}

/// Generation status enum
enum GenerationStatus {
  @JsonValue('pending')
  pending,
  @JsonValue('generating')
  generating,
  @JsonValue('completed')
  completed,
  @JsonValue('failed')
  failed,
  @JsonValue('cancelled')
  cancelled,
}

/// Document type enum
enum DocumentType {
  @JsonValue('resume')
  resume,
  @JsonValue('cover_letter')
  coverLetter,
}

/// Generation progress tracking with stage information
@freezed
class GenerationProgress with _$GenerationProgress {
  const factory GenerationProgress({
    required int currentStage,
    required int totalStages,
    required int percentage,
    String? stageName,
    String? stageDescription,
  }) = _GenerationProgress;

  factory GenerationProgress.fromJson(Map<String, dynamic> json) =>
      _$GenerationProgressFromJson(json);

  /// Initial progress state (queued)
  factory GenerationProgress.initial() => const GenerationProgress(
        currentStage: 0,
        totalStages: 5,
        percentage: 0,
        stageName: null,
        stageDescription: 'Queued for processing',
      );
}

/// Generation result with ATS scores and recommendations
@freezed
class GenerationResult with _$GenerationResult {
  const factory GenerationResult({
    required String documentId,
    required double atsScore,
    required int matchPercentage,
    required double keywordCoverage,
    required int keywordsMatched,
    required int keywordsTotal,
    required String pdfUrl,
    @Default([]) List<String> recommendations,
  }) = _GenerationResult;

  factory GenerationResult.fromJson(Map<String, dynamic> json) =>
      _$GenerationResultFromJson(json);
}

/// Generation options for customizing output
@freezed
class GenerationOptions with _$GenerationOptions {
  const factory GenerationOptions({
    @Default('modern') String template,
    @Default('one_page') String length,
    @Default([]) List<String> focusAreas,
    @Default(false) bool includeCoverLetter,
    String? customInstructions,
  }) = _GenerationOptions;

  factory GenerationOptions.fromJson(Map<String, dynamic> json) =>
      _$GenerationOptionsFromJson(json);

  /// Convert to JSON for API requests
  static Map<String, dynamic> toRequestJson(GenerationOptions options) {
    return {
      'template': options.template,
      'length': options.length,
      'focus_areas': options.focusAreas,
      'include_cover_letter': options.includeCoverLetter,
      if (options.customInstructions != null &&
          options.customInstructions!.isNotEmpty)
        'custom_instructions': options.customInstructions,
    };
  }
}

/// Resume template information
@freezed
class Template with _$Template {
  const factory Template({
    required String id,
    required String name,
    required String description,
    required String previewUrl,
    required List<String> recommendedFor,
    required bool atsFriendly,
  }) = _Template;

  factory Template.fromJson(Map<String, dynamic> json) =>
      _$TemplateFromJson(json);
}

/// Generation list item for simplified list view
@freezed
class GenerationListItem with _$GenerationListItem {
  const factory GenerationListItem({
    required String id,
    required GenerationStatus status,
    required DocumentType documentType,
    required String jobTitle,
    required String company,
    double? atsScore,
    required DateTime createdAt,
    DateTime? completedAt,
  }) = _GenerationListItem;

  factory GenerationListItem.fromJson(Map<String, dynamic> json) =>
      _$GenerationListItemFromJson(json);
}

/// Statistics for generation overview
@freezed
class GenerationStatistics with _$GenerationStatistics {
  const factory GenerationStatistics({
    required int totalGenerations,
    required int completed,
    required int failed,
    required int inProgress,
    required double averageAtsScore,
  }) = _GenerationStatistics;

  factory GenerationStatistics.fromJson(Map<String, dynamic> json) =>
      _$GenerationStatisticsFromJson(json);
}

/// Extension methods for Generation
extension GenerationExtensions on Generation {
  /// Check if generation is complete
  bool get isComplete => status == GenerationStatus.completed;

  /// Check if generation failed
  bool get isFailed => status == GenerationStatus.failed;

  /// Check if generation is currently processing
  bool get isProcessing =>
      status == GenerationStatus.generating ||
      status == GenerationStatus.pending;

  /// Check if generation can be cancelled
  bool get canCancel => isProcessing;

  /// Get display text for status
  String get statusDisplayText {
    switch (status) {
      case GenerationStatus.pending:
        return 'Queued';
      case GenerationStatus.generating:
        return 'Generating...';
      case GenerationStatus.completed:
        return 'Completed';
      case GenerationStatus.failed:
        return 'Failed';
      case GenerationStatus.cancelled:
        return 'Cancelled';
    }
  }

  /// Get color for status
  Color get statusColor {
    switch (status) {
      case GenerationStatus.pending:
      case GenerationStatus.generating:
        return Colors.blue;
      case GenerationStatus.completed:
        return Colors.green;
      case GenerationStatus.failed:
      case GenerationStatus.cancelled:
        return Colors.red;
    }
  }
}

File: backend/app/application/dtos/generation.py
"""Generation API DTOs (Request/Response models)."""

from datetime import datetime
from typing import List, Optional, Literal, Dict, Any
from pydantic import BaseModel, Field, ConfigDict


# Request DTOs
class GenerationOptionsDTO(BaseModel):
    """Generation options request model."""
    template: Literal["modern", "classic", "creative"] = Field(default="modern")
    length: Literal["one_page", "two_page"] = Field(default="one_page")
    focus_areas: List[str] = Field(default_factory=list, max_length=5)
    include_cover_letter: bool = Field(default=False)
    custom_instructions: Optional[str] = Field(None, max_length=500)

    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "template": "modern",
                "length": "one_page",
                "focus_areas": ["backend_development", "leadership"],
                "include_cover_letter": False,
                "custom_instructions": "Emphasize cloud architecture experience"
            }
        }
    )


class GenerateResumeRequest(BaseModel):
    """Request to generate resume."""
    profile_id: str = Field(..., description="Profile ID to use for generation")
    job_id: str = Field(..., description="Job ID to tailor resume for")
    options: Optional[GenerationOptionsDTO] = Field(None, description="Generation options")

    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "profile_id": "550e8400-e29b-41d4-a716-446655440000",
                "job_id": "job-550e8400-e29b-41d4-a716-446655440001",
                "options": {
                    "template": "modern",
                    "length": "one_page"
                }
            }
        }
    )


class GenerateCoverLetterRequest(BaseModel):
    """Request to generate cover letter."""
    profile_id: str = Field(..., description="Profile ID to use for generation")
    job_id: str = Field(..., description="Job ID to tailor cover letter for")
    options: Optional[GenerationOptionsDTO] = Field(None, description="Generation options")


class RegenerateRequest(BaseModel):
    """Request to regenerate with new options."""
    options: Optional[GenerationOptionsDTO] = Field(None, description="Updated generation options")


# Response DTOs
class GenerationProgressDTO(BaseModel):
    """Progress tracking response."""
    current_stage: int = Field(..., ge=0, le=5)
    total_stages: int = Field(default=5)
    percentage: int = Field(..., ge=0, le=100)
    stage_name: Optional[str] = None
    stage_description: Optional[str] = None

    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "current_stage": 2,
                "total_stages": 5,
                "percentage": 40,
                "stage_name": "Profile Compilation",
                "stage_description": "Scoring profile content by relevance"
            }
        }
    )


class GenerationResultDTO(BaseModel):
    """Generation result response."""
    document_id: str
    ats_score: float = Field(..., ge=0.0, le=1.0)
    match_percentage: int = Field(..., ge=0, le=100)
    keyword_coverage: float = Field(..., ge=0.0, le=1.0)
    keywords_matched: Optional[int] = None
    keywords_total: Optional[int] = None
    pdf_url: str
    recommendations: List[str] = Field(default_factory=list)

    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "document_id": "doc-550e8400-e29b-41d4-a716-446655440000",
                "ats_score": 0.87,
                "match_percentage": 82,
                "keyword_coverage": 0.91,
                "keywords_matched": 15,
                "keywords_total": 18,
                "pdf_url": "/api/v1/documents/doc-550e8400-e29b-41d4-a716-446655440000/download",
                "recommendations": [
                    "Add AWS certification to skills",
                    "Quantify team size in leadership experience"
                ]
            }
        }
    )


class GenerationResponse(BaseModel):
    """Generation response model."""
    id: str = Field(..., description="Generation ID (note: 'id' not 'generation_id')")
    status: Literal["pending", "generating", "completed", "failed", "cancelled"]
    progress: GenerationProgressDTO
    profile_id: str
    job_id: str
    document_type: Literal["resume", "cover_letter"] = "resume"
    result: Optional[GenerationResultDTO] = None
    error_message: Optional[str] = None
    tokens_used: int = Field(default=0, ge=0)
    generation_time: Optional[float] = Field(None, ge=0)
    estimated_completion: Optional[str] = None  # ISO datetime string
    created_at: str  # ISO datetime string
    started_at: Optional[str] = None  # ISO datetime string
    completed_at: Optional[str] = None  # ISO datetime string
    updated_at: Optional[str] = None  # ISO datetime string

    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "id": "gen-550e8400-e29b-41d4-a716-446655440000",
                "status": "generating",
                "progress": {
                    "current_stage": 2,
                    "total_stages": 5,
                    "percentage": 40,
                    "stage_name": "Profile Compilation",
                    "stage_description": "Scoring profile content by relevance"
                },
                "profile_id": "550e8400-e29b-41d4-a716-446655440000",
                "job_id": "job-550e8400-e29b-41d4-a716-446655440001",
                "document_type": "resume",
                "tokens_used": 3500,
                "estimated_completion": "2025-11-07T10:30:30Z",
                "created_at": "2025-11-07T10:30:00Z",
                "updated_at": "2025-11-07T10:30:20Z"
            }
        }
    )


class GenerationDetailResultResponse(BaseModel):
    """Detailed generation result with content."""
    id: str
    document_id: str
    document_type: Literal["resume", "cover_letter"]
    content: Dict[str, str] = Field(..., description="Content in multiple formats (text, html, markdown)")
    ats_score: float = Field(..., ge=0.0, le=1.0)
    match_percentage: int = Field(..., ge=0, le=100)
    keyword_coverage: float = Field(..., ge=0.0, le=1.0)
    keywords_matched: int
    keywords_total: int
    pdf_url: str
    recommendations: List[str] = Field(default_factory=list)
    metadata: Dict[str, Any] = Field(..., description="Generation metadata")


class GenerationListItemDTO(BaseModel):
    """Generation list item (summary view)."""
    id: str
    status: Literal["pending", "generating", "completed", "failed", "cancelled"]
    document_type: Literal["resume", "cover_letter"]
    job_title: str
    company: str
    ats_score: Optional[float] = None
    created_at: str  # ISO datetime string
    completed_at: Optional[str] = None  # ISO datetime string


class PaginationDTO(BaseModel):
    """Pagination metadata."""
    total: int
    limit: int
    offset: int
    has_next: bool
    has_previous: bool


class GenerationListResponse(BaseModel):
    """Response for list of generations."""
    generations: List[GenerationListItemDTO]
    pagination: PaginationDTO
    statistics: Dict[str, Any] = Field(default_factory=dict)


class TemplateDTO(BaseModel):
    """Template information."""
    id: str
    name: str
    description: str
    preview_url: str
    recommended_for: List[str]
    ats_friendly: bool


class TemplateListResponse(BaseModel):
    """Response for list of templates."""
    templates: List[TemplateDTO]
